test/indexing.jl

```diff
@@ -1056,7 +1179,163 @@ end
+    @test_throws ArgumentError dfr[Not(3)] = Dict(:a=>10, :c=>11)
+    @test df == DataFrame(a=1,b=2, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws DimensionMismatch dfr[Not(3)] = Dict(:a=>10, :b=>11, :c=>12)
+    @test df == DataFrame(a=1,b=2, c=3)
+
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    dfr[Not(3)] = (a=10, b=11)
+    @test df == DataFrame(a=10,b=11, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws ArgumentError dfr[Not(3)] = (a=10, c=11)
+    @test df == DataFrame(a=1,b=2, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws ArgumentError dfr[Not(3)] = (b=10, a=11)
+    @test df == DataFrame(a=1,b=2, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws DimensionMismatch dfr[Not(3)] = (a=10, b=11, c=12)
+    @test df == DataFrame(a=1,b=2, c=3)
+
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    dfr[Not(3)] = DataFrame(a=10, b=11)[1, :]
+    @test df == DataFrame(a=10,b=11, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws ArgumentError dfr[Not(3)] = DataFrame(a=10, c=11)[1, :]
+    @test df == DataFrame(a=1,b=2, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws ArgumentError dfr[Not(3)] = DataFrame(b=10, a=11)[1, :]
+    @test df == DataFrame(a=1,b=2, c=3)
+    df = DataFrame(a=1,b=2, c=3)
+    dfr = df[1, :]
+    @test_throws DimensionMismatch dfr[Not(3)] = DataFrame(a=10, b=11, c=12)[1, :]
+    @test df == DataFrame(a=1,b=2, c=3)
+end
+
+@testset "setindex! with ! or : and multiple cols" begin
+    df = DataFrame(fill("x", 3, 4))
+    df[!, :] = DataFrame(reshape(1:12, 3, :))
+    @test df == DataFrame(reshape(1:12, 3, :))
+    @test_throws ArgumentError df[!, :] = DataFrame(fill(1, 3, 4))[:, [3,2,1]]
+    @test_throws ArgumentError df[!, :] = DataFrame(fill(1, 3, 4))[1:2, :]
+
+    df = DataFrame(fill("x", 3, 4))
+    df[!, Not(4)] = DataFrame(reshape(1:12, 3, :))[:, 1:3]
+    @test df[:, 1:3] == DataFrame(reshape(1:12, 3, :))[:, 1:3]
+
+    df = DataFrame(fill("x", 3, 4))
+    df[!, :] = reshape(1:12, 3, :)
+    @test df == DataFrame(reshape(1:12, 3, :))
+
+    df = DataFrame(fill("x", 3, 4))
+    df[!, Not(4)] = reshape(1:12, 3, :)[:, 1:3]
+    @test df[:, 1:3] == DataFrame(reshape(1:12, 3, :))[:, 1:3]
+
+    dfv = view(df, :, :)
+    @test_throws ArgumentError dfv[!, :] = DataFrame(reshape(1:12, 3, :))
+    @test_throws ArgumentError dfv[!, :] = reshape(1:12, 3, :)
+
+    for rows in [:, 1:3], cols in [:, r"", Not(r"xx"), 1:4]
+        df = DataFrame(ones(3,4))
+        df[rows, cols] = DataFrame(reshape(1:12, 3, :))
+        @test df == DataFrame(reshape(1:12, 3, :))
+    end
+
+    for rows in [:, 1:3], cols in [:, r"", Not(r"xx"), 1:4]
+        df = DataFrame(ones(3,4))
+        df[rows, cols] = reshape(1:12, 3, :)
+        @test df == DataFrame(reshape(1:12, 3, :))
+    end
+end
+
+@testset "old setindex! tests adjusted to new rules" begin
```

Comment:
This description is a bit weird. Just keep what you think it useful, and drop the old ones.

Reply:
>  Just keep what you think it useful, and drop the old ones.

This is exactly what I did 😄. I renamed it to "additional tests".